R/peaks-functions.R

```diff
@@ -90,6 +90,54 @@ NULL
     x[which(between(x[, "intensity"], intensity)), , drop = FALSE]
 }
 
+#' @description
+#'
+#' Filter the spectrum keeping only peaks that match the provided m/z value(s).
+#'
+#' @inheritParams .peaks_remove
+#'
+#' @param mz `numeric` with the m/z values to match against. This needs to be
+#'     a sorted `numeric` (has to be checked upstream).
+#'
+#' @param tolerance `numeric` with the tolerance. Can be of length 1 or equal
+#'     length `mz`.
+#'
+#' @param ppm `numeric` with the ppm. Can be of length 1 or equal length `mz`.
+#'
+#' @author Johannes Rainer
+#'
+#' @importFrom MsCoreUtils closest
+#'
+#' @noRd
+.peaks_filter_mz_value <- function(x, spectrumMsLevel, mz = numeric(),
+                                   tolerance = 0, ppm = 20,
+                                   msLevel = spectrumMsLevel, ...) {
+    if (!spectrumMsLevel %in% msLevel || !length(x))
+        return(x)
+    idx <- closest(mz, x[, "mz"], tolerance = tolerance, ppm = ppm,
+                   duplicates = "closest")
+    x[idx[!is.na(idx)], , drop = FALSE]
+}
+
+#' @description
+#'
+#' Filter the spectrum keeping only peaks that are within the provided m/z
+#' range.
+#'
+#' @inheritParams .peaks_remove
+#'
+#' @param mz `numeric(2)` with the lower and upper m/z value.
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.peaks_filter_mz_range <- function(x, spectrumMsLevel, mz = numeric(),
+                                   msLevel = spectrumMsLevel, ...) {
+    if (!spectrumMsLevel %in% msLevel || !length(x))
+        return(x)
+    x[which(between(x[, "mz"], mz)), , drop = FALSE]
```

Comment:
Subsetting with `which` was faster in earlier days but now it just offers a benefit if we subset multiple times using the same index:
```r
set.seed(123)
x <- sample(10, 1e2, TRUE)
microbenchmark::microbenchmark(lgl=x[x > 5], which=x[which(x > 5)])
# Unit: microseconds
#   expr   min    lq    mean median     uq   max neval cld
#    lgl 1.098 1.264 1.50135 1.4205 1.6000 7.629   100  a 
#  which 1.586 2.052 2.46258 2.2380 2.4595 9.091   100   b

x <- sample(10, 1e6, TRUE)
microbenchmark::microbenchmark(lgl=x[x > 5], which=x[which(x > 5)])
# Unit: milliseconds
#   expr      min       lq      mean   median       uq      max neval cld
#    lgl 8.180921 8.447758  9.985675 9.222107 11.01384 17.56540   100   a
#  which 8.958257 9.192004 12.349117 9.915709 11.43943 98.04092   100   a
```
IMHO there was a NEWS entry for R 3.x.x somewhere but I can't find it anymore.

Reply:
My main reason to use `which` is to drop any `NA`s. If we can be sure that `between` never returns `NA` we can change - otherwise I would keep it to be on the save side because I think we have to avoid having peaks with `NA` m/z values.

Reply:
OK, so `between` throws an error if there are `NA`s, so it's safe to drop the `which` here. Done in https://github.com/rformassspectrometry/Spectra/commit/02997e36be9e8b53d4937d53b2788f3b3e24c5da